### DISCUSSION TITLE
The working of the local Maximum Filter

### DISCUSSION THREAD

**Discussion by nealresearch** [r-lidar/lidR#712] [General]
I'm struggling to understand how the Local Maximum Filter (LMF) operates, particularly in relation to detecting treetops in a Canopy Height Model (CHM). The paper that introduces this method describes it as flagging a pixel as a treetop if it represents a local maximum within a defined search window. However, my practical experience with implementing the LMF doesn't align with this explanation.

For instance, I have a CHM with a resolution of 0.5m and I've applied a 1m window size for the LMF. Logically, I'd expect to identify a treetop at every 2-pixel interval, but that's not happening. Instead, it seems to be taking the maximum height value in the vicinity of a tree as the treetop. This is inconsistent with a 1m window size. The issue is even more pronounced with larger trees, where multiple points are flagged as treetops, which is what i expect however medium size trees get flagged with only one treetop.

Can someone clarify whether the LMF uses the pixel at the center of the window and then expands equally on either side? For example, if the window size is 1m, is 0.5m to the right and 0.5m to the left of the central pixel included in the window?

I would greatly appreciate a more detailed explanation of how the LMF works.

Here is the link for the CHM

https://drive.google.com/file/d/1rqyyAlm8zCIgGA_Cl5mJ4Z-UJ0b8xl_o/view?usp=sharing

Here is the code  

```r
library("lidR")
library("raster")
library(sf)

chm <- raster("path")

ttops <- locate_trees(chm, lmf(1, shape="circular"))
ttops_2d <- st_zm(ttops, z = NULL, m = NULL)  # Convert 3D points to 2D
st_crs(ttops_2d) <- 25832
sf::st_write(ttops_2d, "path", overwrite = TRUE)
```

**Comment by Jean-Romain** [comment]
Think about it in 2D and in a continuous curve if it is easier. In the figure below, none of the red points are local maxima for the windows size drawn. In this example, no mater the windows size, if the windows is smaller than the distance between the two maxima you will find two maxima, not one maxima every _windows size_. But for super large windows that encompasses the two maxima, you will find only one maxima. The small one being _hidden_ by the higher one.

![lmf](https://github.com/r-lidar/lidR/assets/3872279/758785be-e07d-4608-86d7-39b4323a26fa)

**Reply by nealresearch** [reply]
Thanks for the reply, If we maintain a constant, small window size for the three points mentioned in the figure, should we expect to see three local maxima on the curve? From the figure, it appears that the pixel is centered and the window is constructed around it. How is the window's length distributed? Is it split equally, with half the window size to the left and half to the right of the central pixel? For example, would a window size of 5 consist of 2.5 pixels on the left, the central pixel, and 2.5 pixels on the right?

**Reply by Jean-Romain** [reply]
In my previous figure only the second red point is a local maximum **with a small enough windows**. Other red points are not local maxima. Why to you think that points at random position on the curve should be local maxima? The size of the windows allow to discriminate more local maxima **if they exist** while big windows size only discriminate major maxima. In the following picture the tiny minor noisy peak can onl be detected with a small windows

![lmf](https://github.com/r-lidar/lidR/assets/3872279/e7bfde94-684d-48fe-9227-1ab45b2d335a)

**Reply by nealresearch** [reply]
I think i understood it. Thanks for the clarification.
